Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1913,71 +1943,155 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
-                     }
-                     return S_OK;
-                 }));
+            Factory()->CreateTransformedGeometry(pathGeometry.Get(), __CGAffineTransformToD2D_F(combinedTransform), &transformedGeometry));
+    }
+
+    HRESULT ret = Draw(_kCGCoordinateModeDeviceSpace, nullptr, [&](CGContextRef context, ID2D1DeviceContext* deviceContext) {
+        if (state.textDrawingMode & kCGTextFill) {
+            deviceContext->SetTransform(__CGAffineTransformToD2D_F(combinedTransform));
+            D2D1_POINT_2F origin{ 0, 0 };
+            for (auto& runData : runs) {
+                if (runData.attributes) {
+                    CGColorRef fontColor = (CGColorRef)CFDictionaryGetValue(runData.attributes, _kCGForegroundColorAttributeName);
+                    if (fontColor) {
+                        CGContextSetFillColorWithColor(context, fontColor);
+                    } else {
+                        CFBooleanRef useContextColor =
+                            (CFBooleanRef)CFDictionaryGetValue(runData.attributes, _kCGForegroundColorFromContextAttributeName);
+                        if (!useContextColor || !CFBooleanGetValue(useContextColor)) {
+                            // Neither given a color nor use current context color, so set the fill color to black
+                            CGContextSetRGBFillColor(context, 0, 0, 0, 1);
+                        }
+                    }
+                }
+
+                if (fastPathTextDrawing) {
+                    CGAffineTransform totalTransform =
+                        CGAffineTransformConcat(textTransform,
+                                                CGAffineTransformTranslate(deviceTransform,
+                                                                           runData.relativePosition.x,
+                                                                           std::round(runData.relativePosition.y)));
+                    deviceContext->SetTransform(__CGAffineTransformToD2D_F(totalTransform));
+                }
+
+                deviceContext->DrawGlyphRun(origin, runData.run, state.fillBrush.Get(), DWRITE_MEASURING_MODE_NATURAL);
+            }
+        }
+
+        if (state.textDrawingMode & kCGTextStroke) {
+            deviceContext->SetTransform(__CGAffineTransformToD2D_F(deviceTransform));
+            RETURN_IF_FAILED(_DrawGeometryInternal(transformedGeometry.Get(), kCGPathStroke, context, deviceContext));
+        }
+
+        return S_OK;
+    });
+    RETURN_IF_FAILED(ret);
+
+    if (state.textDrawingMode & kCGTextClip) {
+        RETURN_IF_FAILED(state.IntersectClippingGeometry(transformedGeometry.Get(), kCGPathFill));
+    }
+
+    if (!fastPathTextDrawing) {
+        // Need to delete all of the created glyphOffsets and DWRITE_GLYPH_RUNS
+        for (auto& runData : runs) {
+            delete[] runData.run->glyphOffsets;
```

Comment:
Ideally, but I'm not sure there is one that could be created in a smaller scope than it's needed in, especially since the DWRITE_GLYPH_RUN doesn't have a RAII container for the glyphOffset